### PR TITLE
MF-1314 - Add value comparison filters for readers

### LINKED ISSUE #1314: Consider adding additional filters by SenML `value`
**ENHANCEMENT**
1. Describe the enhancement you are requesting. Enhancements include:
   - tests
   - code refactor
   - documentation
   - research
   - tooling

https://github.com/mainflux/mainflux/pull/1312 introduced filtering by SenML values.

It would probably be interesting to add a few more filters around `value`:
- Less than (`<some_value>`)
- Higher than (`<some_value>`)
- In range (less than `<value_max>` but higher than `<value_min>`)
- Not in range (less than `<value_min>` but higher than `<value_max>`)

2. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
Nice-to-have

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Resolves #1314

### CODE REVIEW

File: readers/messages.go [hunk 1]
```diff
@@ -28,17 +28,40 @@ type MessagesPage struct {
...
+
+// ParseValueComparison convert comparison types to the mathematic anotation
+func ParseValueComparison(query map[string]interface{}) string {
+	comparison := "="
+	val, ok := query["comparison"]
+	if ok {
+		switch val.(string) {
+		case "equal":
+			comparison = "="
+		case "lower-than":
```
Comment (drasko) [inline readers/messages.go]: One question - do we need `lower-than`, or `lower` and `lower-equal` (or maybe `lower-or-equal`) would work?

@dusanb94 @nmarcetic @mteodor @darkodraskovic opinions please?

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/messages.go (hunk 1 above)
Comment (dborovcanin) [reply readers/messages.go]: Or maybe `lt`, `gt`, `le`, `ge`. While it sounds like abbreviations, it is common terminology.

Comment (drasko) [reply readers/messages.go]: You can borrow the naming from assembly conditional instructions: https://developer.arm.com/documentation/dui0068/b/arm-instruction-reference/conditional-execution

File: readers/cassandra/messages_test.go [hunk 2]
```diff
@@ -215,7 +215,72 @@ func TestReadSenml(t *testing.T) {
...
+				Messages: fromSenml(valueMsgs[0:limit]),
+			},
+		},
+		"read message with value and lower-or-equal-than comparison": {
+			chanID: chanID,
+			pageMeta: readers.PageMetadata{
+				Offset:     0,
+				Limit:      limit,
+				Value:      v + 1,
+				Comparison: "lte",
```
Comment (drasko) [inline readers/cassandra/messages_test.go]: I think we mentioned `le` instead of `lte`, according to ARM assembly spec.

@dusanb94 @nmarcetic opinions?

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/openapi.yml [hunk 3]
```diff
@@ -174,6 +174,20 @@ components:
...
+      name: comparison
+      description: Value comparison operator.
+      in: query
+      schema:
+        type: string
+        default: equal
+        enum:
+          - eq
+          - lt
+          - lte
```
Comment (dborovcanin) [inline readers/openapi.yml]: It's `le` and `ge`.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/openapi.yml (hunk 3 above)
Comment (manuio) [reply readers/openapi.yml]: good catch, thank you!

**Review by dborovcanin — DISMISSED** [review]

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**